src/sentry/debug_files/artifact_bundles.py

```diff
@@ -0,0 +1,146 @@
+from collections import defaultdict
+from datetime import datetime
+from typing import DefaultDict, Dict, List, Set
+
+from django.db import IntegrityError, router
+from django.db.models import Q
+
+from sentry.models.artifactbundle import (
+    ArtifactBundle,
+    ArtifactBundleArchive,
+    ArtifactBundleIndex,
+    ArtifactBundleIndexingState,
+)
+from sentry.utils import metrics
+from sentry.utils.db import atomic_transaction
+
+
+def index_artifact_bundles_for_release(
+    artifact_bundles: List[ArtifactBundle], release: str, dist: str
+) -> None:
+    """
+    This indexes the contents of `artifact_bundles` into the database, using
+    the given `release` and `dist` pair.
+
+    Synchronization is achieved using the 4-state `indexing_state`.
+    It is flipped from `needs_indexing` to `is_being_indexed` to avoid conflicts in
+    case two different jobs are calling this method at the same time with the
+    same `artifact_bundle`.
+    Once indexing is done, the state is set to `was_indexed`.
+
+    To avoid too much load and queries on the database, the provided
+    `artifact_bundles` are first merged/deduplicated into an in-memory structure.
+
+    Then, indexing of individual files happens in the following way:
+    - If a key already exists with a higher `date_last_modified`, it is ignored
+      as that indicates out-of-order calls of this method. We always want the
+      most up-to-date bundle to win.
+    - Otherwise, the key is created or updated to point to the `artifact_bundle`.
+    """
+
+    # First: Mark the bundles as `is_being_indexed` using a compare-and-swap like update,
```

Comment:
This probably all should be wrapped in `transaction.atomic` so that a crash in the function (or worker exit) does not leave the bundles in `IS_BEING_INDEXED` state.

Reply:
We would need this mechanism for both state changes, we didn't implement it since we thought about not using a transaction there to speed things up and then using a background job to cleanup but I agree it would be better to guard.

The issue we will still have though, is, if there are crashes during the in-memory merged index we will have to either assume the state will not be changed or have to guard it even more.